oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java

```diff
@@ -1246,6 +1349,256 @@ private void collectRevisionsOlderThan24hAndBetweenCheckpoints(final NodeDocumen
+            }
+            // remember that getAllBundledProperties returns unescaped keys
+            for (String propName : traversedMainNode.getAllBundledProperties().keySet()) {
+                final int lastSlash = propName.lastIndexOf("/");
+                if (lastSlash == -1) {
+                    // then it is an unbundled property which was already handled in phase 1
+                    continue;
+                }
+                final String escapedPropName = Utils.escapePropertyName(propName);
+                // bundled values are of format sub/tree/propertyKey
+                // to extract this we need the last index of "/"
+                final String unbundledSubtreeName = propName.substring(0, lastSlash);
+                final String unbundledPropName = propName.substring(lastSlash + 1);
+                final String unbundledPath = traversedMainNode.getPath().toString() + "/" + unbundledSubtreeName;
+                final DocumentNodeState traversedNode = bundledNodeStates.get(Path.fromString(unbundledPath));
+                if (traversedNode == null) {
+                    log.error("collectUnusedPropertyRevisions : could not find traversed node for bundled key {} unbundledPath {} in doc {}",
+                            propName, unbundledPath, doc.getId());
+                    continue;
+                }
+                final PropertyState traversedProperty = traversedNode.getProperty(unbundledPropName);
+                if (traversedProperty == null) {
+                    log.error("collectUnusedPropertyRevisions : could not get property {} from traversed node {}",
+                            unbundledPropName, traversedNode.getPath());
+                    continue;
+                }
+                deletedRevsCount += removeUnusedPropertyEntries(doc, traversedNode, updateOp, escapedPropName,
+                        (r) -> { updateOp.removeMapEntry(escapedPropName, r); return null;},
+                        allKeepRevs);
+            }
+
+            // phase 3 : "_deleted"
+            int numDeleted = removeUnusedPropertyEntries(doc, traversedMainNode,
+                    updateOp, NodeDocument.DELETED,
+                    (r) -> { NodeDocument.removeDeleted(updateOp, r); return null;} ,
+                    allKeepRevs);
+            deletedRevsCount += numDeleted;
+            return deletedRevsCount;
+        }
+
+        private int collectUnusedInternalPropertyRevisions(final NodeDocument doc,
+                final GCPhases phases, final UpdateOp updateOp,
+                final DocumentNodeState traversedMainNode,
+                final Set<Revision> toKeepUserPropRevs,
+                int deletedRevsCount) {
+            boolean hasUnmergedBranchCommits = false;
+            for (Revision localBc : doc.getLocalBranchCommits()) {
+                if (!isCommitted(nodeStore.getCommitValue(localBc, doc))) {
+                    hasUnmergedBranchCommits = true;
+                }
+            }
```

Comment:
I believe here, we need to have `break` in case `hasUnmergedBranchCommits` is set to true otherwise it might change to false.

```suggestion
            boolean hasUnmergedBranchCommits = doc.getLocalBranchCommits().stream().anyMatch(r -> !isCommitted(nodeStore.getCommitValue(r, doc)));
```

Reply:
indeed, that would have been wrong, nice find!